File: app/src/main/java/br/com/jonathanzanella/myexpenses/card/CardAdapter.java
package br.com.jonathanzanella.myexpenses.card;

import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.annotation.WorkerThread;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import br.com.jonathanzanella.myexpenses.MyApplication;
import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper;
import butterknife.Bind;
import butterknife.ButterKnife;

public class CardAdapter extends RecyclerView.Adapter<CardAdapter.ViewHolder> {
	protected List<Card> cards;
	private CardAdapterCallback callback;

	private final CardRepository cardRepository;
	private ExpenseRepository expenseRepository;

	class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@Bind(R.id.row_card_name)
		TextView name;
		@Bind(R.id.row_card_account)
		TextView account;

		private final AdapterColorHelper adapterColorHelper;
		private final String credit;
		private final String debit;

		public ViewHolder(View itemView) {
			super(itemView);

			ButterKnife.bind(this, itemView);

			credit = itemView.getContext().getString(R.string.credit);
			debit = itemView.getContext().getString(R.string.debit);

			//noinspection deprecation
			int oddColor = itemView.getContext().getResources().getColor(R.color.color_list_odd);
			//noinspection deprecation
			int evenColor = itemView.getContext().getResources().getColor(R.color.color_list_even);
			adapterColorHelper = new AdapterColorHelper(oddColor, evenColor);

			itemView.setOnClickListener(this);
		}

		@UiThread
		public void setData(final Card card) {
			itemView.setBackgroundColor(adapterColorHelper.getColorForLinearLayout(getAdapterPosition()));
			new AsyncTask<Void, Void, Account>() {

				@Override
				protected Account doInBackground(Void... voids) {
					return card.getAccount();
				}

				@Override
				protected void onPostExecute(Account a) {
					super.onPostExecute(a);
					account.setText(a.getName());
				}
			}.execute();

			String cardName = card.getName() + " - ";

			switch (card.getType()) {
				case CREDIT:
					cardName += credit;
					break;
				case DEBIT:
					cardName += debit;
					break;
			}

			name.setText(cardName);
		}

		@Override
		public void onClick(View v) {
			Card card = getCreditCard(getAdapterPosition());
			if(card != null) {
				if(callback != null) {
					callback.onCard(card);
				} else {
					Intent i = new Intent(itemView.getContext(), ShowCardActivity.class);
					i.putExtra(ShowCardActivity.KEY_CREDIT_CARD_UUID, card.getUuid());
					itemView.getContext().startActivity(i);
				}
			}
		}
	}

	public CardAdapter() {
		cardRepository = new CardRepository(new RepositoryImpl<Card>(MyApplication.getContext()), getExpenseRepository());
	}

	private ExpenseRepository getExpenseRepository() {
		if(expenseRepository == null)
			expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(MyApplication.getContext()));
		return expenseRepository;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_card, parent, false);
		return new ViewHolder(v);
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		holder.setData(cards.get(position));
	}

	@Override
	public int getItemCount() {
		return cards != null ? cards.size() : 0;
	}

	@WorkerThread
	public void loadData() {
		cards = cardRepository.all();
	}

	@Nullable
	private Card getCreditCard(int position) {
		return cards != null ? cards.get(position) : null;
	}

	public void setCallback(CardAdapterCallback callback) {
		this.callback = callback;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/card/CardAdapter.kt
package br.com.jonathanzanella.myexpenses.card

import android.content.Intent
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.annotation.WorkerThread
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.MyApplication
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl
import br.com.jonathanzanella.myexpenses.expense.Expense
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper
import br.com.jonathanzanella.myexpenses.views.anko.applyTemplateViewStyles
import br.com.jonathanzanella.myexpenses.views.anko.static
import br.com.jonathanzanella.myexpenses.views.anko.staticWithData
import org.jetbrains.anko.*

class CardAdapter : RecyclerView.Adapter<CardAdapter.ViewHolder>() {
    private var cards: List<Card> = ArrayList()
    private var callback: CardAdapterCallback? = null

    private val cardRepository: CardRepository
    private val expenseRepository = ExpenseRepository(RepositoryImpl<Expense>(MyApplication.getContext()))

    inner class ViewHolder(itemView: View, val ui: CardAdapterViewUI) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val adapterColorHelper: AdapterColorHelper
        private val credit: String = itemView.context.getString(R.string.credit)
        private val debit: String = itemView.context.getString(R.string.debit)

        init {
            val oddColor = ContextCompat.getColor(itemView.context, R.color.color_list_odd)
            val evenColor = ContextCompat.getColor(itemView.context, R.color.color_list_even)
            adapterColorHelper = AdapterColorHelper(oddColor, evenColor)

            itemView.setOnClickListener(this)
        }

        @UiThread
        fun setData(card: Card) {
            itemView.setBackgroundColor(adapterColorHelper.getColorForLinearLayout(adapterPosition))
            object : AsyncTask<Void, Void, Account>() {

                override fun doInBackground(vararg voids: Void): Account {
                    return card.account!!
                }

                override fun onPostExecute(a: Account) {
                    super.onPostExecute(a)
                    ui.cardAccount.text = a.name
                }
            }.execute()

            var cardName = card.name + " - "

            when (card.type) {
                CardType.CREDIT -> cardName += credit
                CardType.DEBIT -> cardName += debit
            }

            ui.cardName.text = cardName
        }

        override fun onClick(v: View) {
            val card = getCreditCard(adapterPosition)
            if (callback != null) {
                callback!!.onCard(card)
            } else {
                val i = Intent(itemView.context, ShowCardActivity::class.java)
                i.putExtra(ShowCardActivity.KEY_CREDIT_CARD_UUID, card.uuid)
                itemView.context.startActivity(i)
            }
        }
    }

    init {
        cardRepository = CardRepository(expenseRepository)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val ui = CardAdapterViewUI()
        return ViewHolder(ui.createView(AnkoContext.create(parent.context, parent)), ui)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(cards[position])
    }

    override fun getItemCount(): Int {
        return cards.size
    }

    @WorkerThread
    fun loadData() {
        cards = cardRepository.all()
    }

    private fun getCreditCard(position: Int): Card {
        return cards[position]
    }

    fun setCallback(callback: CardAdapterCallback) {
        this.callback = callback
    }
}

class CardAdapterViewUI: AnkoComponent<ViewGroup> {
    lateinit var cardName: TextView
    lateinit var cardAccount: TextView

    override fun createView(ui: AnkoContext<ViewGroup>) = with(ui) {
        tableLayout {
            padding = resources.getDimensionPixelSize(R.dimen.default_spacing)
            orientation = TableLayout.HORIZONTAL

            tableRow {
                static { text = resources.getString(R.string.name) }
                cardName = staticWithData { id = R.id.row_card_name }
            }
            tableRow {
                static { text = resources.getString(R.string.account) }
                cardAccount = staticWithData { id = R.id.row_card_account }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
